[GSoC contribution] Improving functions `descendants` and `ancestors` in dag.py

vdshk opened the pull request:
Hi, @dschult!

## Docstring improvements
- [x] Improved `Parameters` section of docstrings
- [x] Added `Raises` section
- [x] Added `Examples` section
- [x] Added `See also` section

## Code improvements
- [x] Rewrote functions to use the faster function `bfs_edges`

rossbar commented on networkx/algorithms/tests/test_dag.py:
```diff
@@ -270,6 +270,10 @@ def test_ancestors(self):
         assert ancestors(G, 1) == set()
         pytest.raises(nx.NetworkXError, ancestors, G, 8)
 
+    def test_ancestors_not_implemented(self):
```
```suggestion
    def test_ancestors_undirected(self):
```
Very minor nit, but a perhaps slightly more meaningful name for the test. Just a suggestion, feel free to ignore - but if you do accept it be sure to change the name for `test_descendants_not_implemented` too so that they're consistent!

vdshk replied:
I fully agree with such names

rossbar reviewed (dismissed):
Very nice improvements @vdshk 

I agree that adding `not_implemented_for('undirected')` makes sense for these functions - AFAICT the names of the functions themselves (as well as the fact that they are in the **d**ag module) imply that they shouldn't be used with undirected graphs.

The only other thing I noticed is the parameter name change in `ancestors`. This is certainly correct and should be changed. In principle, this may cause problems for people who are calling this with `source` as a named kwarg, e.g `ancestors(G, source=0)`, and their could be an argument for adding a FutureWarning to notify users of the name change. IMO though, that's more trouble than it's worth in this case since it's most likely to be called as a positional argument anyways. I'm +1 for the parameter change as proposed here, but LMK if you think it's better to go the warning route instead!

Aside from these points, big +1 for the other improvements!

vdshk commented:
Thanks for the great comments, @rossbar!

I think it's worth discussing how to write FutureWarnings, because both parameter renaming in function `ancestors` and exception change in functions `descendants` and `ancestors` are IMO reasonable but backward incompatible.

rossbar commented:
> I'm a little surprised by the idea that these don't apply to undirected graphs. It seems to me that they are well defined and work just fine for undirected graphs. Is there a reason to keep people from using them that way? I generally prefer not to restrict application of a function unless using it in that context will do harm or cause confusion.

Agreed, and I was hoping to hear what others thought on this subject. My thought was that the naming of the functions was a bit confusing in the context of undirected graphs. For example, `descendants` and `ancestors` will be the same for undirected path graphs, whereas the interpretation of the output makes more sense in the context of directed graphs:

```python
>>> G = nx.path_graph(5)  # An undirected path graph
>>> nx.ancestors(G, 2)
{0, 1, 3, 4}
>>> nx.descendants(G, 2)
{0, 1, 3, 4}
>>> DG = nx.path_graph(5, create_using=nx.DiGraph)
>>> nx.ancestors(DG, 2)
{0, 1}
>>> nx.descendants(DG, 2)
{3, 4}
```

More than anything I think it's just a terminological issue: descendants and ancestors felt like they implied a directionality to me, especially since they are defined in the `dag.py` module.

dschult commented:
Yes -- I agree that they were made for DiGraphs and were implemented first in the special case of DAGs. 
But they work for any DiGraph and in fact work for any Graph. (I can't come up with a name that provides the undirected version of this -- maybe `connected_component`?) In any case, if it is works correctly and is well-defined for the Graph case and doesn't cause any confusion by doing so, I think we should let people use it this way. Otherwise we either direct them to the right function name, or automatically call that other function...  but might as well just run this function. :}

rossbar commented:
> In any case, if it works correctly and is well-defined for the Graph case and doesn't cause any confusion by doing so, I think we should let people use it this way

+1 - in this case, would it be worth moving where these functions are defined, similar to #5016? It should be relatively easy to raise a FutureWarning for users who are accessing these functions directly from the `dag` module, and for the vast majority of users who call them from the main namespace (e.g. `nx.ancestors`) there will be no difference. OTOH maybe this is too much churn for too little benefit... treating this minor organizational point as a "wontfix" also seems sensible.

In any case, this PR still has a lot of nice improvements - @vdshk would you mind reverting the changes related to `not_implemented_for`?

vdshk commented:
+1 for that these functions work correctly and are well-defined for the Graph. Fixed this mistake.

I also rewrote these functions to use the faster function `bfs_edges` as part of other improvements.

ericmjl commented:
I just had a chance to catch up with @rossbar and @vdshk today about the issue.

@dschult I kinda think that there _might_ be initial confusion with the behavior of the function with undirected graphs. This arises from a mismatch between an intuitive understanding of the terms vs. a rigorous understanding of the behavior of the functions. 

One might be able to make the mathematical case that tracing ancestors of a node requires recursively tracing all in-edges that bring one to that node, and that all undirected graphs can be re-interpreted as bidirectional edge directed graphs, and hence in this case the behavior is well-defined... and hence it will yield the equivalent of connected components. However, knowing that undirected graphs can be reinterpreted as a special case of directed graphs might not be common knowledge.

What I think is a reasonable prior on common knowledge is that a user of ancestors and descendants will expect this to work with directed graphs. Being permissive with the function's use on undirected graphs may be surprising for the average user.

Having typed out the logic that tries to explain why the function should work with undirected graphs, I think there are the following ways to make the function less confusing:

1. Include a concise version of the aforementioned explanation in the docstring so that the behavior with undirected graphs is unsurprising and well-explained, and move the function out of DAGs.
2. Restrict the function such that it only works with DAGs, i.e. not just DiGraphs but acyclic ones, such that calling the function on anything non-DiGraph and non-acyclic will yield an error. (This would be a breaking change, so we'd do a warning first, as @rossbar suggested.)

In our chat with @vdshk today, we brought up the phenomena that NetworkX has gradually become the _de facto_ library for graph analysis especially amongst Pythonistas. I think we've got a good opportunity to educate here! A function might be well-defined but still confusing behaviorally. In this case, I think we could do more to provide clarity.

@dschult, what do you think would be the more appropriate course of action?

dschult commented:
Thanks @ericmjl for this summary. I agree that it would be helpful to have a sentence or two in the documentation about how an undirected graph will be handled by allowing both directions for each edge. That is quite pervasive in this library. It even goes as deep as the choice to use 2-tuples to represent undirected edges instead of frozen-sets.  It's quite possible we made a mistake in that choice, but that is another discussion. I think most people can fairly quickly understand this idea if it is explained.  An undirected edge is treated as allowing connections in either direction. (or similar).

Of the two options, I prefer option 1) where we add some to the doc_string and move these functions out of `dag.py`.  The moving out of `dag.py` is much less important to me. I think we should all be using the `nx.` namespace. People only started using the full path to the code when our documentation started using the full path to the function names and we said we would look for a way to avoid that -- but it hasn't happened yet. Perhaps we need to start forcing people to use the main namespace, but that seems a little strict too. :}

If we move it, perhaps we can provide an alias for it in the old location (not advertised in the docs, but available for backward compatibility).  @rossbar is there any precedence for doing something like that?

dschult reviewed (approved):
Revisiting the docstrings as they are.... Nothing in this implies that it only works for undirected graphs. 
Once the input parameter is labeled as "A NetworkX Graph" the one line description says it all.  I don't think we need anything other than "Returns all nodes having a path to `source` in `G`".  
Am I missing something?

So, I think this PR is ready to merge.  We can create another one to move these functions out of `dag.py` is desired.

rossbar commented:
> If we move it, perhaps we can provide an alias for it in the old location (not advertised in the docs, but available for backward compatibility). @rossbar is there any precedence for doing something like that?

In my head it would look something like adding a `__get_attr__` to `dag.py` that raises a warning when someone imports/accesses the functions directly from dag.py. That way, when someone does something like:

```python
>>> from networkx.algorithms import dag
>>> dag.ancestors(G, 0)
```
or
```python
>>> from networkx.algorithms.dag import ancestors
```
They'll get some sort of `FutureWarning` stating that these functions will be moved to a new module in 3.0. The nice thing is that users accessing the functions from the main namespace (which should be most users) won't see any difference and won't get any warnings.

I agree that it's not very important, but *I think* the `__get_attr__` gives us the tools we need to clear this up without being too noisy. Either way, it should definitely be addressed elsewhere!

rossbar reviewed (approved):
LGTM, thanks @vdshk for identifying this and improving the docstrings!
